staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go

```diff
@@ -322,101 +322,103 @@ func unescapeSingleQuote(s string) (string, error) {
-			}
-			fieldName := s.TokenText()
-			if isMapSyntax {
-				if tok == scanner.Char {
-					newS := fieldName[1 : len(fieldName)-1]
-					newS, err := unescapeSingleQuote(newS)
-					if err != nil {
-						return pathOfFieldPath, field.Invalid(pathOfFieldPath, fieldPath, fmt.Sprintf("failed to unescape: %v", err))
+	// configure the scanner to split the string into tokens.
+	// The three delimiters ('.', '[', ']') will be returned as single char tokens.
+	// All other text between delimiters is returned as string tokens.
+	scanner.Split(func(data []byte, atEOF bool) (advance int, token []byte, err error) {
+		if len(data) > 0 {
+			for i := 0; i < len(data); i++ {
+				// If in a single quoted string, look for the end of string
+				// ignoring delimiters.
+				if data[0] == '\'' {
+					if i > 0 && data[i] == '\'' && data[i-1] != '\\' {
+						// Return quoted string
+						return i + 1, data[:i+1], nil
 					}
-					fieldName = newS
-					if schema.AdditionalProperties != nil {
-						validFieldPath = validFieldPath.Key(fieldName)
+					continue
+				}
+				switch data[i] {
+				case '.', '[', ']': // delimiters
+					if i == 0 {
+						// Return the delimiter.
+						return 1, data[:1], nil
 					} else {
-						validFieldPath = validFieldPath.Child(fieldName)
+						// Return identifier leading up to the delimiter.
+						// The next call to split will return the delimiter.
+						return i, data[:i], nil
 					}
-				} else {
-					return pathOfFieldPath, field.Invalid(pathOfFieldPath, fieldPath, "unexpected format of fieldName: "+fieldName)
-				}
-			} else if tok != scanner.Ident {
-				return pathOfFieldPath, invalidFieldError
-			} else {
-				if schema.AdditionalProperties != nil {
-					validFieldPath = validFieldPath.Key(fieldName)
-				} else {
-					validFieldPath = validFieldPath.Child(fieldName)
 				}
 			}
-
-			if schema.Properties != nil {
-				propertySchema, ok := schema.Properties[fieldName]
-				if ok {
-					schema = &propertySchema
-				} else {
-					return pathOfFieldPath, invalidFieldError
-				}
-			} else if schema.AdditionalProperties != nil {
-				schema = schema.AdditionalProperties.Structural
-			} else {
-				return pathOfFieldPath, invalidFieldError
+			if atEOF {
+				// Return the string.
+				return len(data), data, nil
 			}
+		}
+		return 0, nil, nil
+	})
 
-			if isMapSyntax {
-				if tok == scanner.EOF {
-					return pathOfFieldPath, field.Invalid(pathOfFieldPath, fieldPath, "unexpected end of JSON path")
-				}
-				s.Scan()
-				if s.TokenText() != "]" {
-					return pathOfFieldPath, field.Invalid(pathOfFieldPath, fieldPath, "expect ] but get: "+s.TokenText())
-				}
+	var tok string
+	for scanner.Scan() {
+		tok = scanner.Text()
+		switch tok {
+		case "[":
+			if !scanner.Scan() {
+				return nil, fmt.Errorf("unexpected end of JSON path")
+			}
+			tok = scanner.Text()
+			if !strings.HasPrefix(tok, "'") || !strings.HasSuffix(tok, "'") {
```

Comment:
nit: Probably should also need to check that string length is at least 2 here, the tokenizer isn't super smart and could return a single quote back.

Reply:
Updated